### PR TITLE
[FIX] base, hr: avoid invalid access error in onchange() when editing user

### PR DESCRIPTION
**Pull request by thle-odoo**
Steps to reproduce:
-------------------
- install the "hr" module;
- remove access rights for "Employees";
- change language on the user profile.

Issue:
------
There's an Access Error because we can't read the `private_street` field
on the employee that corresponds to the user.

Cause:
------
The new version of onchange fetches the record values on the server side,
unlike the old version which used the values in the view.

In the old version, as we were using view values, this didn't cause any problems,
as the values came from a read that which took into account `SELF_READABLE_FIELDS`.

Note:
We do not have access to the value of the `private_street` field because it is a
related field with the attribute `related_sudo=False` and we do not have
access rights for the `hr.employee` model.

Solution:
---------
Use the cache and place the values of the fields in `SELF_READABLE_FIELDS` in it
before performing the onchange logic.

opw-3664929

### CODE REVIEW

File: addons/hr/tests/test_self_user_access.py [hunk 1]
```diff
@@ -216,3 +216,15 @@ def testWriteOtherUserEmployee(self):
...
+
+    def test_onchange_readable_fields_with_no_access(self):
+        """
+            The objectif is to test that the onchange logic takes account of `SELF_READABLE_FIELDS`.
+
+            The view contains fields in `SELF_READABLE_FIELDS` (example: `private_street`).
+            Even if the user does not have read access to the employee,
+            it should not cause an access error if these fields are in `SELF_READABLE_FIELDS`.
+        """
+        self.env.ref("base.lang_fr").active = True
```
Comment (rco-odoo) [inline addons/hr/tests/test_self_user_access.py:228]: There is a dedicated method to ensure that a language is active (mostly used in tests)
```suggestion
        self.env['res.lang']._activate_lang("fr_FR")
```

File: addons/hr/tests/test_self_user_access.py [hunk 2]
```diff
@@ -216,3 +216,15 @@ def testWriteOtherUserEmployee(self):
...
+        user_form = Form(self.richard.with_user(self.richard), view=self.env.ref('hr.res_users_view_form_profile'))
+        user_form.lang = "fr_FR" # Trigger an onchange
```
Comment (rco-odoo) [inline addons/hr/tests/test_self_user_access.py:230]: You can use the `with` form of it, and `view` can be a string
```suggestion
        with Form(self.richard.with_user(self.richard), view='hr.res_users_view_form_profile') as form:
            # triggering an onchange should not trigger some access error
            form.lang = "fr_FR"
```

**Review by rco-odoo — COMMENTED** [review]
Looks reasonably fine.

Please rewrite the commit title to describe the problem you are fixing.
I suggest "[FIX] base, hr: avoid invalid access error in onchange() when editing user".

File: addons/hr/tests/test_self_user_access.py [hunk 3]
```diff
@@ -216,3 +216,15 @@ def testWriteOtherUserEmployee(self):
     def testSearchUserEMployee(self):
         # Searching user based on employee_id field should not raise bad query error
         self.env['res.users'].with_user(self.richard).search([('employee_id', 'ilike', 'Hubert')])
+
+    def test_onchange_readable_fields_with_no_access(self):
+        """
+            The objectif is to test that the onchange logic takes account of `SELF_READABLE_FIELDS`.
+
+            The view contains fields in `SELF_READABLE_FIELDS` (example: `private_street`).
```
Comment (rco-odoo) [inline addons/hr/tests/test_self_user_access.py:224]:
```suggestion
            The purpose is to test that the onchange logic takes into account `SELF_READABLE_FIELDS`.

            The view contains fields that are in `SELF_READABLE_FIELDS` (example: `private_street`).
```

**Review by ryv-odoo — APPROVED** [review]

**Comment by rco-odoo** [comment]
@robodoo rebase-ff r+

**Pull request closed without merging**